src/dmd/e2ir.d

```diff
@@ -3531,6 +3531,18 @@ elem *toElem(Expression e, IRState *irs)
 
         override void visit(CondExp ce)
         {
+            if (auto ve = ce.econd.isVarExp)
+                if (ve.var.ident == Id.ctfe)
+                {
+                    elem *e = toElem(ce.e2, irs);
+                    if (irs.params.cov && ce.e2.loc.linnum)
+                        e = el_combine(incUsageElem(irs, ce.e2.loc), e);
+                    if (tybasic(e.Ety) == TYstruct)
+                        e.ET = Type_toCtype(ce.e2.type);
+                    elem_setLoc(e, ce.loc);
+                    result = e;
+                    return;
+                }
```

Comment:
Seems like a good idea to me. Needs to be in a different PR though and I guess there's already a bug report for this somewhere.

Ideally, this should not hard-detect the ctfe variable though. You code will fail on constructs such as these:
```d
void foo(bool dummy)
{
    if (__ctfe && dummy)
    {
        // CTFE-only Code goes here
    }
}
```
What should be done instead is this: Set `__ctfe` (somehow) to false, then constant fold the condition. If the result is a constant `false`, the if part of the condition is a ctfe only part. If it is constant true, the `else` branch is a ctfe only branch. If it is not const, both branches may be used at runtime.

Maybe that's out of scope for GSoC though, but it would be very useful for betterC.

Reply:
I wonder how all that stuff integrates with SCOPEctfe though, maybe what should really happen is that the if block content should be set to SCOPEctfe instead and the toElem / nogc visitors should skip these scopes.

OTOH there's no scope introduced when using CondExp, so for CondExp your code here seems to be corret.

Reply:
ping @JinShil as IIRC your working on SCOPEctfe issues, so you may be interested in this.

Reply:
What is this code doing and why?

Reply:
It detects an `__ctfe ? e1 : e2` expression, then ignores `e1` and only sends `e2` to the code generation stage. I guess without this, dropping `e1` code occurs later in the backend. @Vild can you elaborate why exactly this is necessary?

Reply:
Without this change the codegen would generate code for both `e1` and `e2`, and while generating code for `e1` it would trigger this `assert(0)` https://github.com/dlang/dmd/pull/9982/files#diff-7a340f667e49df4f91bcab621efa8521R3294

It will also only trigger on code that looks like `__ctfe ? e1 : e2`, as this is the only form I'm outputting inside `expressionsem.d`. But I probably need to extend this to atleast support `if` statements as the inline can rewrite `CondExp` to `IfStatement`.

Reply:
> Without this change the codegen would generate code for both e1 and e2, and while generating code for e1 it would trigger this assert(0).

OK. But somehow the backend currently also drops the `e1` branch, the code is not in the generated object file. Does this happen only as dead-code-elimination after the initial codegen stage?

Reply:
> Does this happen only as dead-code-elimination after the initial codegen stage?

I have been unable to find where in the backend were it gets removed, all I know is that it is removed.

Reply:
> It detects an `__ctfe ? e1 : e2` expression, then ignores `e1` and only sends `e2` to the code generation stage.

So, this is basically a workaround because we don't have `version(CTFE)`?

> Without this change the codegen would generate code for both `e1` and `e2`, and while generating code for `e1` it would trigger this `assert(0)`.

Why wasn't `e1` lowered in the semantic phase?

Reply:
> So, this is basically a workaround because we don't have version(CTFE)?
ya

> Why wasn't e1 lowered in the semantic phase?
Because `e1` is the `__ctfe` code-path, which is the original expression.